deep_qa/training/trainer.py

```diff
@@ -509,6 +515,95 @@ def _dataset_indexing_kwargs(self) -> Dict[str, Any]:
     # Protected methods - you CAN override these, if you want
     ###################
 
+    def _compile_parallel_model(self) -> DeepQaModel:
+        """
+        This method compiles a multi-gpu version of your model. This is done using data
+        parallelism, by making N copies of the model on the different GPU, all of which
+        share parameters. Gradients are updated synchronously, using the average gradient
+        from all of the outputs of the various models. This effectively allows you to scale
+        a model up to batch_sizes which cannot fit on a single GPU.
+
+        This method returns a "primary" copy of the model, which has had it's training
+        function which is run by Keras overridden to be a training function which trains
+        all of the towers of the model. The other towers never have their training functions
+        initialised or used and are completely hidden from the user. The returned model
+        can be serialised in the same way as any other model and has no dependency on
+        multiple gpus being available when it is loaded.
+
+        Returns
+        -------
+        The "primary" copy of the DeepQaModel, which holds the training function which
+        trains all of the copies of the model.
+        """
+        tower_models = []
+        tower_gradients = []
+        global_step = tensorflow.train.get_or_create_global_step()
+        train_loss = tensorflow.get_variable('train_loss', [],
+                                             initializer=tensorflow.constant_initializer(0.0),
+                                             trainable=False)
+
+        # Place a copy of the model on each GPU, each getting a slice of the batch.
+        for gpu_index in range(self.num_gpus):
+            with tensorflow.device(pin_variable_device_scope('/gpu:%d' % gpu_index)):
+                with tensorflow.name_scope('tower_%d' % gpu_index):
+                    # This is a new model object every time.
+                    model = self._build_model()
```

Comment:
Does `_build_model` ever have any side effects?  Like creating objects and stashing them away for re-use?